jax/scipy/optimize/bfgs_minimize.py

```diff
@@ -0,0 +1,140 @@
+"""The Broyden-Fletcher-Goldfarb-Shanno minimization algorithm.
+https://pages.mtu.edu/~struther/Courses/OLD/Sp2013/5630/Jorge_Nocedal_Numerical_optimization_267490.pdf
+"""
+
+import jax
+import jax.numpy as jnp
+from jax.lax import while_loop
+from .line_search import line_search
+from typing import NamedTuple, Optional
+
+
+class BFGSResults(NamedTuple):
+  converged: bool  # bool, True if minimization converges
+  failed: bool  # bool, True if line search fails
+  k: int  # The number of iterations of the BFGS update.
+  nfev: int  # The total number of objective evaluations performed.
+  ngev: int  # total number of jacobian evaluations
+  nhev: int  # total number of hessian evaluations
+  x_k: jnp.ndarray  # A tensor containing the last argument value found during the search. If the search converged,
+  # then this value is the argmin of the objective function.
+  f_k: jnp.ndarray  # A tensor containing the value of the objective
+  # function at the `position`. If the search
+  # converged, then this is the (local) minimum of
+  # the objective function.
+  g_k: jnp.ndarray  # A tensor containing the gradient of the objective function at the `final_position`.
+  # If the search converged the l2-norm of this tensor should be below the tolerance.
+  H_k: jnp.ndarray  # A tensor containing the inverse of the estimated Hessian.
+
+
+def bfgs_minimize(func, x0, options=None, _nojit=False):
+  """
+  The BFGS algorithm from
+      Algorithm 6.1 from Wright and Nocedal, 'Numerical Optimization', 1999, pg. 136-143
+
+      Notes:
+          We utilise boolean arithmetic to avoid jax.cond calls which don't work on accelerators.
+          A side effect is that we perform more gradient evaluations than scipy's BFGS
```

Comment:
To clarify, `jax.lax.cond` is supported on accelerators. It is relatively more expensive than `jax.numpy.where` and/or small amounts of arithmetic, but still probably cheaper than evaluating the objective function/gradient evaluations. We should still use it for that case.

Reply:
Ah, from your other [comment](https://github.com/google/jax/pull/3101#discussion_r425527173) above I had thought that it was more like all `cond` had to be carried out on a CPU. Back in the day GPU kernels didn't handle if statements. Can `scan`, and `while_loop` be evaluated on an accelerator? What precisely are the performance dings, so that I can make a better decision about what to use?

Reply:
All these ops are supported when executing programs on accelerators. For GPUs, the control flow itself happens back on the CPU, handled transparently by XLA.

This is fine if a considerable amount of work is being done or avoided, which is true for evaluating expensive functions like neural networks, but not for small amounts of math. For example, in https://github.com/google/jax/pull/3076 the overhead was roughly 25 us per loop iteration.

25 us is quite slow compared to GPU arithmetic even on large arrays, but is quite fast compared to most evaluations of neural networks (typically ms).

Reply:
Okay, well, what I can do it make a second version with cond's and then profile them. I don't have a GPU around for profiling on, though I could use colab. Is that a good option?

Reply:
Colab is fine for profiling, but a representative test suite seems challenging. For now, I think it would be best just to stick to a simple rule, e.g., "Use `where` instead of `cond`, unless doing so would increase the number of function or gradient evaluations"

Reply:
Thanks, for this PR, I'll leave as is, and assign this also to the project of implementing DCSRCH as a primary line search.

Reply:
I read through the DCSRCH [code](https://github.com/optimizers/nlpy/tree/master/nlpy/optimize/ls/src) and find that there is no real benefit to implementing it. It also use a mixture of cubic, quadratic, and bisection for stepsize zooming except with more conditionals.

Reply:
To be clear fmin_bfgs tries dcsrch and if it fails then tries the other method.